py/mattak/backends/uproot/dataset.py

```diff
@@ -364,28 +378,32 @@ def wfs(self, calibrated : bool = False, channels : Optional[Union[int, List[int
                 starting_window = self._get_windows(dict(entry_start=idx, entry_stop=idx+1, library='np'))
         else:
             # so ... we need to loop through and find which things we have actually have waveforms
-            # start by allocating the output
-            w = numpy.zeros((self.last - self.first, self.NUM_CHANNELS, self.NUM_WF_SAMPLES), dtype='float64' if calibrated else 'int16')
-            starting_window = numpy.zeros((self.last - self.first))
             # now figure out how much of the data array we need
             wf_start = None
             wf_end = None
 
             # store the indices that will be non-zero
             wf_idxs = []
-            for i in range(self.first, self.last):
-                if i in self.events_with_waveforms:
-                    wf_idxs.append(i - self.first)
+            for idx in range(self.first, self.last):
+                if idx in self.events_with_waveforms:
+                    wf_idxs.append(idx - self.first)
                     # these are the start and stop of our array we need to load
                     if wf_start is None:
-                        wf_start = self.events_with_waveforms[i]
+                        wf_start = self.events_with_waveforms[idx]
 
-                    wf_end = self.events_with_waveforms[i] + 1
+                    wf_end = self.events_with_waveforms[idx] + 1
 
             if len(wf_idxs):
-                w[wf_idxs] = self._get_waveforms(dict(entry_start=wf_start, entry_stop=wf_end, library='np'))
-                starting_window[wf_idxs] = self._get_windows(dict(entry_start=wf_start, entry_stop=wf_end, library='np'))
-
+                # take the overwrite if it is not None
+                skip_incomplete = overwrite_skip_incomplete or self.skip_incomplete
+                if skip_incomplete:
+                    w = self._get_waveforms(dict(entry_start=wf_start, entry_stop=wf_end, library='np'))
+                    starting_window = self._get_windows(dict(entry_start=wf_start, entry_stop=wf_end, library='np'))
+                else:
+                    w = numpy.empty((self.last - self.first, self.NUM_CHANNELS, self.NUM_WF_SAMPLES), dtype='float64' if calibrated else 'int16')
+                    starting_window = numpy.empty((self.last - self.first, self.NUM_CHANNELS, 2))
```

Comment:
is numpy.empty better than numpy.zeroes? Sure, it skips zeroing the the array for events that are present, but then you have uninitialized data in your array that could maybe cause confusion?

Reply:
exactly I would not use `empty` because it is harder to spot.